src/hooks/preupdate_hook.rs

```diff
@@ -0,0 +1,333 @@
+    }
+
+    /// Get the depth of the query that triggered the preupdate hook.
+    /// Returns 0 if the preupdate callback was invoked as a result of
+    /// a direct insert, update, or delete operation;
+    /// 1 for inserts, updates, or deletes invoked by top-level triggers;
+    /// 2 for changes resulting from triggers called by top-level triggers; and so forth.
+    pub fn get_query_depth(&self) -> i32 {
+        unsafe { ffi::sqlite3_preupdate_depth(self.db) }
+    }
+
+    /// Get the row id of the row being updated/deleted.
+    pub fn get_old_row_id(&self) -> i64 {
+        self.old_row_id
+    }
+
+    /// Get the value of the row being updated/deleted at the specified index.
+    pub fn get_old_column_value(&self, i: i32) -> ValueRef {
+        let mut p_value: *mut ffi::sqlite3_value = ptr::null_mut();
+        unsafe {
+            ffi::sqlite3_preupdate_old(self.db, i, &mut p_value);
+            ValueRef::from_value(p_value)
+        }
+    }
+}
+
+/// An accessor to access the new values of the row being inserted/updated
+/// during the preupdate callback.
+#[derive(Debug)]
+pub struct PreUpdateNewValueAccessor {
+    db: *mut ffi::sqlite3,
+    new_row_id: i64,
+}
+
+impl PreUpdateNewValueAccessor {
+    /// Get the amount of columns in the row being inserted/updated.
+    pub fn get_column_count(&self) -> i32 {
+        unsafe { ffi::sqlite3_preupdate_count(self.db) }
+    }
+
+    /// Get the depth of the query that triggered the preupdate hook.
+    /// Returns 0 if the preupdate callback was invoked as a result of
+    /// a direct insert, update, or delete operation;
+    /// 1 for inserts, updates, or deletes invoked by top-level triggers;
+    /// 2 for changes resulting from triggers called by top-level triggers; and so forth.
+    pub fn get_query_depth(&self) -> i32 {
+        unsafe { ffi::sqlite3_preupdate_depth(self.db) }
+    }
+
+    /// Get the row id of the row being inserted/updated.
+    pub fn get_new_row_id(&self) -> i64 {
+        self.new_row_id
+    }
+
+    /// Get the value of the row being updated/deleted at the specified index.
+    pub fn get_new_column_value(&self, i: i32) -> ValueRef {
+        let mut p_value: *mut ffi::sqlite3_value = ptr::null_mut();
+        unsafe {
+            ffi::sqlite3_preupdate_new(self.db, i, &mut p_value);
+            ValueRef::from_value(p_value)
+        }
+    }
+}
+
+impl Connection {
+    /// Register a callback function to be invoked before
+    /// a row is updated, inserted or deleted.
+    ///
+    /// The callback parameters are:
+    ///
+    /// - the name of the database ("main", "temp", ...),
+    /// - the name of the table that is updated,
+    /// - a variant of the PreUpdateCase enum which allows access to extra functions depending
+    /// on whether it's an update, delete or insert.
+    #[inline]
+    pub fn preupdate_hook<'c, F>(&'c self, hook: Option<F>)
+    where
+        F: FnMut(Action, &str, &str, &PreUpdateCase) + Send + 'c,
```

Comment:
The `'c` (instead of `'static`) seems wrong to me but I am not sure.

Reply:
Yes, you're correct. This was a remnant from the original PR where all the other hooks had lifetimes attached. Missed that when I was updating this, thanks!